solsys_code/solsys_code_observatory/views.py

```diff
@@ -0,0 +1,151 @@
+            obs.short_name = self.obs_data['short_name']
+            if self.obs_data['old_names']:
+                obs.old_names = self.obs_data['old_names']
+            elong = float(self.obs_data['longitude'])
+            obs.lon = elong
+            # Convert parallax constants to longitude (again), latitude and altitude
+            obs.from_parallax_constants(elong, float(self.obs_data['rhocosphi']), float(self.obs_data['rhosinphi']))
+            try:
+                created_time = datetime.strptime(self.obs_data['created_at'], '%a, %d %b %Y %H:%M:%S %Z')
+                created_time = created_time.replace(tzinfo=timezone.utc)
+                obs.created = created_time
+            except ValueError:
+                pass
+            try:
+                modified_time = datetime.strptime(self.obs_data['updated_at'], '%a, %d %b %Y %H:%M:%S %Z')
+                modified_time = modified_time.replace(tzinfo=timezone.utc)
+                obs.modified = modified_time
+            except ValueError:
+                pass
+            # Make dictionary of choices using dict comprehension
+            obstype_choices = {choice[1].lower(): choice[0] for choice in Observatory.OBSTYPE_CHOICES}
+            obs.observations_type = obstype_choices.get(self.obs_data['observations_type'], 0)
+            obs.uses_two_line_obs = self.obs_data['uses_two_line_observations']
+
+            obs.save()
+        return obs
+
+
+class CreateObservatory(CreateView):
+    """Creates an Observatory object by retrieving the MPC code from the form
+    and querying the MPC Obscodes API
+    (https://www.minorplanetcenter.net/mpcops/documentation/obscodes-api/)"""
+
+    model = Observatory
+    form_class = CreateObservatoryForm
+    template_name = 'solsys_code_observatory/observatory_create.html'
+
+    def get_success_url(self):
+        """Create a custom success_url to redirect to the detail page for the
+        newly created Observatory.
+        """
+        return reverse_lazy('solsys_code_observatory:detail', kwargs={'pk': self.kwargs['pk']})
+
+    def get_context_data(self, **kwargs):  # noqa: D102
+        context = super().get_context_data(**kwargs)
+        return context
+
+    def form_valid(self, form: BaseModelForm) -> HttpResponse:
+        """
+        Runs after the form validation (which ensures that the obscode is uppercased (required))
+        Performs the query through the MPC API using ``MPCObscodeFetcher()`` and then tries to
+        create the ``Observatory`` through ``MPCObscodeFetcher.to_observatory(). This means
+        we shouldn't/don't call the superclass's ``form_valid`` method as this will
+        attempt to create a duplicate (which raises an IntegrityError)
+        """
+        obs = MPCObscodeFetcher()
+        obs.query(form.cleaned_data['obscode'])
+        try:
+            obs = obs.to_observatory()
+            self.object = obs
+            self.kwargs['pk'] = obs.pk
+        except IntegrityError:
```

Comment:
If you are going to include the `MissingDataException` above, you should catch that here as well and pass a messages.error in the same way. 

```
except MissingDataException as e:
    messages.error(self.request, f'No observatory returned: {e}'
```